plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsWriter.java

```diff
@@ -194,51 +193,53 @@ private CollectedStatistics mergeStatisticsIfNecessary(
-                                    .filter(blobMetadata -> blobMetadata.type().equals(APACHE_DATASKETCHES_THETA_V1))
-                                    .filter(blobMetadata -> blobMetadata.inputFields().size() == 1)
-                                    .filter(blobMetadata -> pendingPreviousNdvSketches.contains(getOnlyElement(blobMetadata.inputFields())))
-                                    .collect(toImmutableList());
-                            for (Pair<BlobMetadata, ByteBuffer> read : reader.readAll(toRead)) {
-                                Integer fieldId = getOnlyElement(read.first().inputFields());
-                                checkState(pendingPreviousNdvSketches.remove(fieldId), "Unwanted read of stats for field %s", fieldId);
-                                Memory memory = Memory.wrap(ByteBuffers.getBytes(read.second())); // Memory.wrap(ByteBuffer) results in a different deserialized state
-                                CompactSketch previousSketch = CompactSketch.wrap(memory);
-                                CompactSketch newSketch = requireNonNull(collectedNdvSketches.get(fieldId), "ndvSketches.get(fieldId) is null");
-                                ndvSketches.put(fieldId, SetOperation.builder().buildUnion().union(previousSketch, newSketch));
+            if (hasUsefulData) {
+                try (PuffinReader reader = Puffin.read(fileIO.newInputFile(statisticsFile.path()))
+                        .withFileSize(statisticsFile.fileSizeInBytes())
+                        .withFooterSize(statisticsFile.fileFooterSizeInBytes())
+                        .build()) {
+                    List<BlobMetadata> toRead = reader.fileMetadata().blobs().stream()
+                            .filter(blobMetadata -> blobMetadata.type().equals(APACHE_DATASKETCHES_THETA_V1))
+                            .filter(blobMetadata -> blobMetadata.inputFields().size() == 1)
+                            .collect(toImmutableList());
+                    for (Pair<BlobMetadata, ByteBuffer> read : reader.readAll(toRead)) {
+                        Integer fieldId = getOnlyElement(read.first().inputFields());
+                        if (columnsWithRecentlyComputedStats.remove(fieldId)) {
+                            CompactSketch newSketch = requireNonNull(collectedNdvSketches.get(fieldId), "ndvSketches.get(fieldId) is null");
+                            if (finalUpdateMode == INCREMENTAL_UPDATE) {
+                                newSketch = SetOperation.builder().buildUnion().union(readPreviousSketch(read), newSketch);
                             }
+                            ndvSketches.put(fieldId, newSketch);
                         }
-                        catch (IOException exception) {
-                            throw new UncheckedIOException(exception);
+                        else {
+                            ndvSketches.put(fieldId, readPreviousSketch(read));
```

Comment:
This is when new stats contain no information about a column, but the old file has.
In such case, we don't need to deserialize the sketch, we can copy it as-is.
More importantly though -- where is the code copying non-Theta blobs from the old Puffin file to the new one?